src/crypto/tls/ech.go

```diff
@@ -0,0 +1,818 @@
+
+		// Ensure that only TLS 1.3 and above are offered.
+		if v := helloInner.supportedVersions; v[len(v)-1] < VersionTLS13 {
+			return nil, nil, errors.New("tls: ech: invalid version for ClientHelloInner")
+		}
+
+		// Set "random".
+		if c.ech.hrrInnerRandom == nil {
+			// Generate a fresh "random".
+			helloInner.random = make([]byte, 32)
+			_, err := io.ReadFull(config.rand(), helloInner.random)
+			if err != nil {
+				return nil, nil, errors.New("tls: short read from Rand: " + err.Error())
+			}
+		} else {
+			// After HRR, use the "random" sent in the first ClientHelloInner.
+			helloInner.random = c.ech.hrrInnerRandom
+		}
+
+		// Prepare the ClientHelloOuter. Generate a fresh ClientHello, ignoring
+		// the generated key share. This will be replaced with the key share
+		// generated for helloBase.
+		hello, _, err = c.makeClientHello(config.MinVersion)
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: ech: %s", err)
+		}
+
+		// Set "random".
+		hello.random = helloBase.random
+
+		// Set "legacy_session_id" to be the same as ClientHelloInner.
+		hello.sessionId = helloBase.sessionId
+
+		// Set "key_share" to the same as ClientHelloInner.
+		hello.keyShares = helloBase.keyShares
+
+		// Set "server_name" to be the client-facing server.
+		hello.serverName = hostnameInSNI(string(echConfig.rawPublicName))
+
+		// Prepare the encryption context. Note that c.ech.hrrPsk is initially
+		// nil, meaning no PSK is used for encrypting the first ClientHelloInner
+		// in case of HRR.
+		ctx, enc, err := echConfig.setupClientContext(c.ech.hrrPsk, config.rand())
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: ech: %s", err)
+		}
+
+		// Compress the ClientHelloInner using the "outer_extension" mechanism,
+		// incorporating the "key_share" extension from ClientHelloOuter.
+		//
+		// NOTE(cjpatton): It would be nice to incorporate more extensions, but
+		// "key_share" is the last extension to appear in the ClientHello before
+		// "pre_shared_key". As a result, the only contiguous sequence of outer
+		// extensions that contains "key_share" is "key_share" itself. Note that
+		// we cannot change the order of extensions in the ClientHello, as the
+		// unit tests expect "key_share" to be second to last extension.
+		innerDigest := ctx.innerDigest(helloInner.marshal())
+		outerExtensions := []uint16{extensionKeyShare}
+		if testingECHOuterExtMany {
+			// NOTE(cjpatton): Incorporating this particular sequence does not
+			// yield significant savings. However, it's useful to test that our
+			// server correctly handles a sequence of compressed extensions and
+			// not just one.
+			outerExtensions = []uint16{
+				extensionStatusRequest,
+				extensionSupportedCurves,
+				extensionSupportedPoints,
+			}
+		}
+		rawCompressedHello, ok := echIncorporateOuterExtensions(helloInner.marshal(), innerDigest, outerExtensions)
+		if !ok {
+			return nil, nil, errors.New("tls: ech: compression of ClientHelloInner failed")
+		}
```

Comment:
I left the padding alone, since it's still in flux. In particular, if https://github.com/tlswg/draft-ietf-tls-esni/pull/313 lands in -08, then this changes the existing client-side padding.

Reply:
👍